### PR TITLE
Journal user events

### PR DESCRIPTION
**Pull request by tiago-loureiro**
Allow to optionally journal the following events over protobuf to SQS:
 * User activation
 * User deletion
 * User updates (for locale and email)

### CODE REVIEW

File: libs/types-common-journal/proto/UserEvents.proto [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    enum EventType {
+        USER_ACTIVATE     = 0;
+        USER_DELETE       = 1;
+        USER_UPDATE       = 2;
+        // --^ When used, the new values for optional fields are
+        //     whatever is set (email, locale) unless they are 
+        //     null / not set (protobuf does not distinguish)
+        USER_EMAIL_REMOVE = 3;
+        // --^ When an email is removed, the email field is set
+        //     to the value of the old email address
```
Comment (jschaul) [inline libs/types-common-journal/proto/UserEvents.proto]: > the email field is set to the value of the old email address

User A registers with email; then adds a phone number, then removes his email (all valid operations w.r.t the API). There would be no "old value" for anyone to set. This comment is not very clear.

File: deploy/services-demo/conf/brig.demo.yaml [hunk 2]
```diff
@@ -27,9 +27,10 @@ gundeck:
 aws:
   sesQueue: integration-brig-events
   internalQueue: integration-brig-events-internal
+  journalQueue: integration-user-events
```
Comment (jschaul) [inline deploy/services-demo/conf/brig.demo.yaml]: journaling is not needed in demo-mode, let's comment it out here.

Comment (tiago-loureiro) [reply deploy/services-demo/conf/brig.demo.yaml]: 👍

File: libs/types-common-journal/proto/UserEvents.proto (hunk 1 above)
Comment (tiago-loureiro) [reply libs/types-common-journal/proto/UserEvents.proto]: In that case, the value of the old email (the one that the user just removed) will be on the event https://github.com/wireapp/wire-server/blob/develop/services/brig/src/Brig/API/User.hs#L416

File: services/brig/test/integration/Util/AWS.hs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+isRealSESEnv env = case view AWS.journalQueue env of
+    Just url | "amazonaws.com" `Text.isSuffixOf` url -> True
+    _                                                -> False
+
+purgeJournalQueue :: AWS.Env -> IO ()
+purgeJournalQueue env = for_ (view AWS.journalQueue env)
+                      $ SQS.execute (view AWS.amazonkaEnv env)
+                      . SQS.purgeQueue
+
+assertEmptyUserQueue :: AWS.Env -> IO ()
```
Comment (fisx) [inline services/brig/test/integration/Util/AWS.hs]: Why `UserQueue`, not `JournalQueue`?

Comment (fisx) [reply services/brig/test/integration/Util/AWS.hs]: (same for `assertUserQueue`.)

File: services/brig/test/integration/Main.hs [hunk 4]
```diff
@@ -64,8 +67,11 @@ runTests iConf bConf otherArgs = do
     lg <- Logger.new Logger.defSettings
     db <- defInitCassandra casKey casHost casPort lg
     mg <- newManager tlsManagerSettings
-
-    userApi     <- User.tests bConf mg b c ch g =<< mkLocalAWSEnv lg awsOpts mg
+    awsEnv <- AWS.mkEnv lg awsOpts mg
+    
+    -- If we're journaling, make sure we have a clean slate
+    for_ (awsEnv^.AWS.journalQueue) $ SQS.execute (view AWS.amazonkaEnv awsEnv) . SQS.purgeQueue
```
Comment (fisx) [inline services/brig/test/integration/Main.hs]: this is happening before each test, too, so it should not be strictly necessary here, no?  but i don't mind if it remains there.

File: libs/types-common-journal/proto/UserEvents.proto (hunk 1 above)
Comment (fisx) [reply libs/types-common-journal/proto/UserEvents.proto]: what about `when an email address X is removed, the email field is set to X`?

File: services/brig/src/Brig/IO/Intra.hs [hunk 5]
```diff
@@ -143,6 +146,15 @@ updateSearchIndex orig e = case e of
...
 
+journalEvent :: UserId -> UserEvent -> AppIO ()
+journalEvent orig e = case e of
+    UserActivated acc                 -> Journal.userActivate (accountUser acc)
+    UserLocaleUpdated _ loc           -> Journal.userUpdate orig Nothing (Just loc)
+    UserIdentityUpdated _ (Just em) _ -> Journal.userUpdate orig (Just em) Nothing
+    UserIdentityRemoved _ (Just em) _ -> Journal.userEmailRemove orig em
+    UserDeleted{}                     -> Journal.userDelete orig
+    _                                 -> return ()
+
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/brig/src/Brig/IO/Intra.hs]: > This means we have email addresses of not-activated users piling up in brig that the journal does not know about.

Yes, I think we should not journal unverified email addresses; once they get verified, we will journal them (this also avoids tracking, on our end, email addresses that are not verified/that you do not own). Note that unverified email/phone number get automatically deleted after [a certain amount of time](https://github.com/wireapp/wire-server/blob/develop/services/brig/src/Brig/Options.hs#L342-L345)

File: libs/types-common-journal/proto/UserEvents.proto (hunk 1 above)
Comment (tiago-loureiro) [reply libs/types-common-journal/proto/UserEvents.proto]: > what about when an email address X is removed, the email field is set to X?

👍

**Pull request merged**